apps/prairielearn/src/pages/instructorAssessmentInstance/instructorAssessmentInstance.ejs

```diff
@@ -318,6 +366,25 @@
             <tr>
               <td class="text-nowrap"><%= row.formatted_date %></td>
               <td><%= row.auth_user_uid %></td>
+              <td><a tabindex="0" class="btn btn-xs btn-secondary"
+                id="fingerprintPopover<%= row.client_fingerprint_id %>"
+                data-toggle="popover" data-container="body"
+                data-html="true" data-placement="auto" title="Fingerprint"
+                data-content="
+                  <div>
+                    IP Address: <%= row.ip_address %>
+                  </div>
+                  <div>
+                    Session ID: <%= row.user_session_id %>
+                  </div>
+                  <div>
+                    User Agent: <%= row.user_agent %>
+                  </div>
+                "
+                data-trigger="focus" onclick="$(this).popover({trigger: focus})">
+               <%= row.client_fingerprint_id %>
```

Comment:
The `client_fingerprint_id` is going to be a horrible number like 76475533. I'd prefer to always count from 1 on each assessment instance. That way if there are two different IPs being used we'd see the fingerprint switch back and forwards between `1` and `2` rather than between `76475533` and `76493818` (for example). As a stretch goal we could even have the badge color depend on the fingerprint number, so `1` is gray, `2` is blue, etc. If we had say 8 colors then we could just cycle through them. This would help instructors see immediately where changes happened.

To implement this I suggest doing it in `assessment.selectAssessmentInstanceLog()` and doing it in JS (not in the SQL). We could add an extra property to `InstanceLogSchema` called something like `client_fingerprint_number` which would be a `z.number().optional()` so it doesn't need to be returned by the SQL. Then in the JS we can iterate through the log entries and add `client_fingerprint_number` whenever we have a `client_fingerprint.id`, using a Map to store the mapping from `client_fingerprint.id` to `client_fingerprint_number`. Let me know if it isn't clear how to use a Map to do this.

Reply:
Maybe this changing-color thing isn't really needed -- let's try it first with having everything be gray and we can think about color-changing later.

Reply:
I have added a map in `assessment.selectAssessmentInstanceLog()` that maps the `client_fingerprint.id` to a new number that increments for each new fingerprint starting at 1. This piece of code was a little bit tricky so some review would be helpful. 

I have not yet added color tags but I do agree that it would be helpful for identifying where the fingerprint changes. Let me know if you want me to have a go at adding color tags where they are displayed.

Reply:
We would like to get this merged soon (like this week) so let's not worry about colors for now, we can do that in a later PR.